### PR TITLE
[Autofocus v2] Capture the API points usage

### PR DESCRIPTION
**Pull request by jlevypaloalto**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: [link to the issue](https://jira-dc.paloaltonetworks.com/browse/CIAC-5777)

## Description
Add API metrics, point usage chart, limit rescheduling to AutoFocus V2.

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/CommonDashboards/ReleaseNotes/1_3_10.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+
+#### Dashboards
+
+##### API Execution Metrics
+
+- Added a widget to provide API execution metrics for the ***AutoFocus V2*** integration.
```
Comment (ShirleyDenkberg) [inline Packs/CommonDashboards/ReleaseNotes/1_3_10.md:6]:
```suggestion
Added a widget to provide API execution metrics for the **AutoFocus V2** integration.
```

File: Packs/AutoFocus/Integrations/AutofocusV2/AutofocusV2.py [hunk 2]
```diff
@@ -2,16 +2,9 @@
...
-import json
-import requests
 import socket
 import traceback
-import urllib3
-from typing import Callable, Tuple
-
-# Disable insecure warnings
-urllib3.disable_warnings()
+from collections.abc import Callable
```
Comment (JudahSchwartz) [inline Packs/AutoFocus/Integrations/AutofocusV2/AutofocusV2.py:7]: changed from typing? is this an issue?

File: Packs/AutoFocus/Integrations/AutofocusV2/AutofocusV2.py [hunk 3]
```diff
@@ -254,24 +247,87 @@
...
+def rerun_command_if_required(api_res: dict, rate_limit_auto_retry: bool):
+    daily_points_remaining = dict_safe_get(api_res, ('bucket_info', 'daily_points_remaining'), 0)
+    if not (rate_limit_auto_retry and daily_points_remaining):
+        results = CommandResults(
+            readable_output=(
+                f'Error in API call to AutoFocus:\nMessage: {api_res.get("message")}\n'),
+            entry_type=EntryType.ERROR
+        )
+    else:
+        next_run_in_seconds = int(dict_safe_get(api_res, ('bucket_info', 'wait_in_seconds'), 70)) + 1  # type: ignore
```
Comment (JudahSchwartz) [inline Packs/AutoFocus/Integrations/AutofocusV2/AutofocusV2.py:303]: what if this number here is very high? waiting until tomorrow? We should check it cant be larger than a certain number (to be set as an advanced integration parameter with a default of five minutes)

Comment (JudahSchwartz) [inline Packs/AutoFocus/Integrations/AutofocusV2/AutofocusV2.py:303]: we need to give a greater buffer than one second. At least 20

File: Packs/AutoFocus/Integrations/AutofocusV2/AutofocusV2.py [hunk 4]
```diff
@@ -254,24 +247,87 @@
...
+def return_metrics():
+    if EXECUTION_METRICS.metrics is not None and ExecutionMetrics.is_supported():
+        return_results(EXECUTION_METRICS.metrics)
+    if API_POINTS_TABLE.readable_output:
+        return_results(API_POINTS_TABLE)
+
+
+def rerun_command_if_required(api_res: dict, rate_limit_auto_retry: bool):
+    daily_points_remaining = dict_safe_get(api_res, ('bucket_info', 'daily_points_remaining'), 0)
+    if not (rate_limit_auto_retry and daily_points_remaining):
```
Comment (JudahSchwartz) [inline Packs/AutoFocus/Integrations/AutofocusV2/AutofocusV2.py:296]: flip this if so we dont need the not

File: Packs/AutoFocus/Integrations/AutofocusV2/AutofocusV2.py [hunk 5]
```diff
@@ -1849,77 +1901,89 @@ def get_export_list_command(args):
...
-            get_export_list_command(args)
-        elif active_command == 'ip':
-            return_results(search_ip_command(**args))
-        elif active_command == 'domain':
-            return_results(search_domain_command(**args))
-        elif active_command == 'url':
-            return_results(search_url_command(**args))
-        elif active_command == 'file':
-            return_results(search_file_command(**args))
+        match command:
```
Comment (JudahSchwartz) [inline Packs/AutoFocus/Integrations/AutofocusV2/AutofocusV2.py:1931]: if at first you dont succeed...

**Review by JudahSchwartz — COMMENTED** [review]
looks great! a couple minor comments, lets tag doc review

File: Packs/AutoFocus/Integrations/AutofocusV2/AutofocusV2.py (hunk 2 above)
Comment (jlevypaloalto) [reply Packs/AutoFocus/Integrations/AutofocusV2/AutofocusV2.py:7]: it's a flake8 thing

File: Packs/AutoFocus/Integrations/AutofocusV2/AutofocusV2.py [hunk 6]
```diff
@@ -291,24 +291,23 @@ def return_metrics():
...
-    if not (rate_limit_auto_retry and daily_points_remaining):
-        results = CommandResults(
-            readable_output=(
-                f'Error in API call to AutoFocus:\nMessage: {api_res.get("message")}\n'),
-            entry_type=EntryType.ERROR
-        )
-    else:
-        next_run_in_seconds = int(dict_safe_get(api_res, ('bucket_info', 'wait_in_seconds'), 70)) + 1  # type: ignore
+    if retry_on_rate_limit and daily_points_remaining:
+        next_run_in_seconds = min(int(dict_safe_get(api_res, ('bucket_info', 'wait_in_seconds'), 70)) + 20, 300)  # type: ignore
```
Comment (JudahSchwartz) [inline Packs/AutoFocus/Integrations/AutofocusV2/AutofocusV2.py:297]: if wait_in_seconds from the api is 2 hours for some reason, we should just not retry at all. In this case we will be polling every five minutes for no reason.
should have a 
if next_run_in_seconds > 10 minutes as part of this if above

**Review by JudahSchwartz — APPROVED** [review]
looks good! last comment. Also note that we didnt technically in force a 1 poll only policy, but i think this is fine

**Comment by jlevypaloalto** [comment]
This PR needs a force merge as Autofocus has flaky TPBs.

**Pull request merged**